### PR TITLE
Unify costing APIs in WaterTAP

### LINKED ISSUE #676: WaterTAP costing improvements
There are a few potential improvements that I've been thinking about for WaterTAP costing:

- [x] Moving the creation of *global* unit costing parameters (typically fixed variables) from the build_global_params in the costing package to the unit costing methods (Done in #835 for WaterTAPCosting Package, ~~TODO for ZeroOrderCosting Package~~) (Working on #911 I realized the ZO Costing package has *always* done this.)

The benefits would include:
- Avoiding the creation of the parameters if the unit is not costed on the flowsheet, currently all of the costing parameters get added regardless of whether the unit exists. The issue is that when someone displays the costing parameter block there are many unused parameters and is confusing for the user - this will only grow as we add more units.
- The unit costing method could create a block with the unit name to the costing package parameter block, this would help with the name space and organization of the parameters

- [ ] Moving the unit costing methods to their unit models (Started in #835 for WaterTAPCosting Package, DONE in #911 for ZeroOrderCosting Package, Completed in #1122)

The benefits would include:
- Clearer separation of the system costing and the unit costing, this also helps with documentation. The unit costing could have their own technical reference page.
- Making it easier for users to see what is required for creating a custom unit costing because each unit costing file is essentially an example.
- It would make it more clear to users if we have multiple costing methods for a unit, they could see multiple costing files and import them as needed as opposed to configuration arguments that I think is more convoluted

- [ ]  Documentation (e.g., #723) (Completed in #1122, moved to #1135)
- [ ]  #710 (Completed in #1122)
- [x] Allow user-settable `base_currency`, and hard-code all parameters to use a fixed year. (Done in #835)
- [x] Rename `Var` / `Param`s for flow cost (e.g., `electricity_base_cost`) to `{flow_type}_cost` (e.g., `electricity_cost`) using IDAES/idaes-pse#983; (#865)
- [x] Consider combining the two costing packages into a single package (moved to #1134)

### LINKED ISSUE #710: Add expression for total annualized cost to WT costing packages
Suggested by @bknueven in #709. The expression to be added is:

`costing.total_annualized_cost = costing.total_capital_cost * costing.capital_recovery_factor + m.fs.costing.total_operating_cost`

This is used in the default definition of LCOW and it is frequently used in other levelized cost metrics as well.

### PR DESCRIPTION
**Pull request by bknueven**
## Summary/Motivation:
See #676.

## Changes proposed in this PR:

- [x] Creates a simple base class for costing, which extends the IDAES Flowsheet Costing Block
- [x] Moves duplicated functionality into the new base class
- [x] Adds documentation for this base class
- [x] Adds documentation for WaterTAPCosting Package and Utilities.
- [x] Both costing packages inherit from this base class
- [x] Removes some custom methods and tracking from the `WaterTAPCosting` package from when everything was constructed as a monolith.
- [x] Misc. updates to the code to keep up with these changes.
- [x] Add tests for using ZO models with WaterTAP Costing Package and high-fidelity models with the Zero-Order Costing Package

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by bknueven** [comment]
Let's wait for #1120 before merging this -- I would like to address this comment as well: https://github.com/watertap-org/watertap/pull/1120#issuecomment-1710603283

File: watertap/costing/watertap_costing_package.py [hunk 1]
```diff
@@ -10,141 +10,31 @@
...
-
+class WaterTAPCostingData(WaterTAPCostingBlockData):
     def build_global_params(self):
         # Register currency and conversion rates based on CE Index
         register_idaes_currency_units()
 
+        self._build_common_global_params()
+
         # Set the base year for all costs
         self.base_currency = pyo.units.USD_2018
```
Comment (kurbansitterley) [inline watertap/costing/watertap_costing_package.py:32]: While we are making major changes to costing, can we change our default currency to be something more recent? 2021 is available as a unit, but probably not a good representative/default costing year since the pandemic made year/year inflation high, but 2020 or 2022 might be better. But maybe changing default parameter values is something you want to do in a separate PR.

File: watertap/costing/unit_models/ion_exchange.py [hunk 2]
```diff
@@ -34,7 +34,7 @@ def build_hcl_cost_param_block(blk):
     )
 
     costing = blk.parent_block()
-    costing.add_defined_flow("HCl", blk.cost / blk.purity)
+    costing.register_flow_type("HCl", blk.cost / blk.purity)
```
Comment (kurbansitterley) [inline watertap/costing/unit_models/ion_exchange.py:37]: This is unique to IX maybe, but is it possible to only register one of these material flows based on a CONFIG argument for the IX model? The model only requires one of these for costing, so having all just adds bloat

File: watertap/costing/watertap_costing_package.py (hunk 1 above)
Comment (bknueven) [reply watertap/costing/watertap_costing_package.py:32]: Right now nothing in this PR changes the underlying models in any way, it's just a giant reshuffling and consolidation of existing code and functionality.

I would suggest putting off a change like this for when we merge the costing packages.

File: watertap/costing/unit_models/ion_exchange.py (hunk 2 above)
Comment (bknueven) [reply watertap/costing/unit_models/ion_exchange.py:37]: Not in a straightforward fashion, but of course anything can be done.

Right now the decorators ensure that the parameters are declared exactly once, even if two or more of the same unit is costed in a flowsheet.

The costing method doesn't need to use the decorators, however, to register the parameters. It could do it manually while ensuring they only get added once.

File: watertap/costing/watertap_costing_package.py [hunk 3]
```diff
@@ -160,99 +50,24 @@ def build_global_params(self):
...
-        )
-        self.add_defined_flow("electricity", self.electricity_cost)
-
-        self.electrical_carbon_intensity = pyo.Param(
-            mutable=True,
-            initialize=0.475,
-            doc="Grid carbon intensity [kgCO2_eq/kWh]",
-            units=pyo.units.kg / pyo.units.kWh,
-        )
+        self.capital_recovery_factor.expr = self.factor_capital_annualization
```
Comment (kurbansitterley) [inline watertap/costing/watertap_costing_package.py:53]: This seems not ideal but I imagine will be rectified once we agree on how we want to handle which parameters to include and how they will be calculated/what their default values will be.

File: watertap/costing/costing_base.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        """
+        Allow the unit model to register its default costing method,
+        either through an attribute named "default_costing_method"
+        or by naming the default costing method "default_costing_method"
+        """
+        if hasattr(unit_model, "default_costing_method"):
+            return unit_model.default_costing_method
+        return super()._get_costing_method_for(unit_model)
+
+    def register_flow_type(self, flow_type, cost):
```
Comment (kurbansitterley) [inline watertap/costing/costing_base.py:230]: Can you explain what this is doing and why it is necessary? No issue with the code itself, just what it is doing in the larger framework here and why are units involved.

**Review by kurbansitterley — APPROVED** [review]
LGTM. I'd like to see further consolidation of costing packages, but that is for a future PR.

File: watertap/costing/watertap_costing_package.py (hunk 3 above)
Comment (bknueven) [reply watertap/costing/watertap_costing_package.py:53]: Yeah, I agree it is a bit ugly. But it lets us put `add_LCOW` and similar aggregates in the base class.

File: watertap/costing/costing_base.py (hunk 4 above)
Comment (bknueven) [reply watertap/costing/costing_base.py:230]: Ah, this goes back to something I think you found a while ago.

The IDAES costing framework will create a new `Var` in `register_flow_type`, unless there already is a component on the costing package named `flow_type + "_cost"`.

But this creates issues if you want to use mutable parameters or other variables to define the `cost`, then changes to those won't be reflected. There's a bit more discussion on the issue we're side-stepping here: https://github.com/IDAES/idaes-pse/pull/1014#issuecomment-1309527271.

Comment (bknueven) [reply watertap/costing/costing_base.py:230]: As for the units, it has to do with the fact that Pyomo treats units as expressions. So, e.g., we want `register_flow_type( "foo", 42.0 * pyo.units.USD_2020 / pyo.units.m )` to go ahead and create a new variable called `foo_cost` whose value is `42.0 * pyo.units.USD_2020 / pyo.units.m`.

But if I define a parameter or variable named `bar_base_cost` and another parameter called `bar_purity`, then `register_flow_type( "bar", bar_base_cost / bar_purity)`, we will *not* create a new `Var`, but instead create a new `Expression` called `bar_cost` with `bar_cost.expr = bar_base_cost / bar_purity`. It is assumed that if the user wants to do parameter sweeps over the bar_cost the user will modify `bar_base_cost`.


There is a proposed fix for some of this ugliness, but it hasn't been a priority for IDAES/Pyomo: https://github.com/IDAES/idaes-pse/pull/1014#issuecomment-1332709400

File: watertap/costing/unit_models/electrodialysis.py [hunk 5]
```diff
@@ -55,7 +55,10 @@ def cost_electrodialysis(blk, cost_electricity_flow=True, has_rectifier=False):
     Function for costing the Electrodialysis unit
 
     Args:
-        cost_electricity_flow - Option for including the costing of electricity
+        cost_electricity_flow (:obj:`bool`, optional): Option for including the
+            costing of electricity. Defaults to True.
+        has_retifier (:obj:`bool`, optional): Option for including a rectifier.
```
Comment (adam-a-a) [inline watertap/costing/unit_models/electrodialysis.py]:
```suggestion
        has_rectifier (:obj:`bool`, optional): Option for including a rectifier.
```

File: watertap/core/util/misc.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    Returns:
+        bool : True if the expr is constant besides units, False otherwise
+    """
+    for v in identify_variables(expr):
+        return False
+    for p in identify_mutable_parameters(expr):
+        if isinstance(p, _PyomoUnit):
+            continue
+        return False
+    return True
```
Comment (adam-a-a) [inline watertap/core/util/misc.py:31]: Can you explain the logic behind this function? If variables are identified in an expression, or if any mutable params in the expression don't have units, we deem it as not being constant--that's what I'm interpreting but still not sure and also not absorbing the purpose.

File: watertap/core/util/tests/test_misc.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+    # test non-mutable param with unit-only expression
+    assert is_constant_up_to_units(m.fixed_param * units.m**2 / units.s)
+
+    m.fixed_param_2 = Param(initialize=6.28)
+
+    # test expression of fixed params with units
+    assert is_constant_up_to_units(
+        (m.fixed_param**2) * m.fixed_param_2 * units.m**2
+    )
```
Comment (adam-a-a) [inline watertap/core/util/tests/test_misc.py:46]: OK so this returns True because it is an immutable Param, and if it were mutable, it would return False -- so it seems.

File: watertap/core/util/tests/test_misc.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+    m.mutable_param = Param(initialize=42, mutable=True)
+
+    # test mutable param
+    assert not is_constant_up_to_units(m.mutable_param)
+
+    m.mutable_param_units = Param(initialize=42, units=units.m**2 / units.s)
+    # test mutable param with units specified
+    assert not is_constant_up_to_units(m.mutable_param_units)
```
Comment (adam-a-a) [inline watertap/core/util/tests/test_misc.py:55]: Does Pyomo automatically designate a Param as mutable when units are assigned? If not, then I don't get this.

File: watertap/costing/costing_base.py [hunk 9]
```diff
@@ -35,6 +35,188 @@ class WaterTAPCostingBlockData(FlowsheetCostingBlockData):
...
+                pyo.units.convert(
+                    flow_rate, to_units=pyo.units.m**3 / self.base_period
+                )
+                * self.utilization_factor
+            ),
+            doc=f"Constraint for Levelized Cost of Water based on flow {flow_rate.name}",
+        )
+        self.add_component(name + "_constraint", LCOW_constraint)
+
+        self._registered_LCOWs[name] = (LCOW, LCOW_constraint)
```
Comment (adam-a-a) [inline watertap/costing/costing_base.py:73]: I don't think it would work with the existing formulation, but I think we eventually want to be able to get the LCOW breakdown, which would yield levelized cost components for capex and opex across a flowsheet. Just something to think about because it would be extremely useful to have at our fingertips. I manually did this in the LSRRO flowsheet when we working on that analysis.

File: watertap/costing/costing_base.py [hunk 10]
```diff
@@ -35,6 +35,188 @@ class WaterTAPCostingBlockData(FlowsheetCostingBlockData):
...
+
+    def add_specific_energy_consumption(
```
Comment (adam-a-a) [inline watertap/costing/costing_base.py:75]: SImilarly, it'd be useful to get a full SEC breakdown by component in addition to this aggregate SEC for a model

File: watertap/costing/costing_base.py [hunk 11]
```diff
@@ -35,6 +35,188 @@ class WaterTAPCostingBlockData(FlowsheetCostingBlockData):
...
+        )
+
+    def add_electricity_intensity(self, flow_rate, name="electricity_intensity"):
+        """
+        Add calculation of overall electricity intensity to costing block.
+        Args:
+            flow_rate - flow rate of water (volumetric) to be used in
+                        calculating electricity intensity
+            name (optional) - the name of the Expression for the specific
+                              electrical intensity (default: specific_electrical_carbon_intensity)
```
Comment (adam-a-a) [inline watertap/costing/costing_base.py:127]: I was never a big fan of the "electricity intensity" usage for the WT3 migration, and I think as we move towards consolidating into one package, we should reduce to specific energy consumption (which technically could be with respect to electricity or thermal energy, so we'd have to consider that as well).

File: watertap/costing/costing_base.py [hunk 12]
```diff
@@ -35,6 +35,188 @@ class WaterTAPCostingBlockData(FlowsheetCostingBlockData):
...
+        """
+        self.add_specific_energy_consumption(flow_rate, name=name)
+
+    def add_specific_electrical_carbon_intensity(
+        self, flow_rate, name="specific_electrical_carbon_intensity"
+    ):
+        """
```
Comment (adam-a-a) [inline watertap/costing/costing_base.py:134]: As we consolidate packages, we may want to consider using wordage that aligns with the national decarbonization effort (e.g., CO2_equivalent_footprint or whatever is ubiquitously used nowadays).

File: watertap/costing/costing_base.py [hunk 13]
```diff
@@ -35,6 +35,188 @@ class WaterTAPCostingBlockData(FlowsheetCostingBlockData):
...
+            doc="Total capital cost of the process",
+            units=self.base_currency,
+        )
+        self.total_operating_cost = pyo.Var(
+            initialize=0,
+            doc="Total operating cost of process per operating period",
+            units=self.base_currency / self.base_period,
+        )
+
+    def _build_common_global_params(self):
```
Comment (adam-a-a) [inline watertap/costing/costing_base.py:174]: So replaced `practical_investment_factor` with  `TIC`? I'd be good with that.

File: watertap/costing/watertap_costing_package.py [hunk 14]
```diff
@@ -96,6 +96,14 @@ def build_process_costs(self):
                 + sum(self.aggregate_flow_costs.values()) * self.utilization_factor
             )
 
+        self.total_annualized_cost = pyo.Expression(
+            expr=(
+                self.total_capital_cost * self.capital_recovery_factor
```
Comment (adam-a-a) [inline watertap/costing/watertap_costing_package.py:101]: Similar to my earlier comments for future work, with regard to a breakdown of LCOW cost components, would be great to have an `annualized_cost` attribute for every capital costs. With that, all we'd need to do is then divide by the desired flowrate.

File: docs/technical_reference/costing/costing_base.rst [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+Specific Energy Consumption
++++++++++++++++++++++++++++
+
+For a given volumetric flow `Q`, the specific energy consumption, :math:`E_{spec,Q}` is calculated by the `add_specific_energy_consumption` method as
+
+    .. math::
+    
+        E_{spec,Q} = \frac{C_{el,tot}}{Q}
+
+Specific Electrical Carbon Intesity
```
Comment (adam-a-a) [inline docs/technical_reference/costing/costing_base.rst]: "Intensity"

File: docs/technical_reference/costing/costing_base.rst [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+=============================================  ====================  =====================================  ===============  ==============================================================================
+Plant capacity utilization factor                 :math:`f_{util}`    ``utilization_factor``                 90%                Percentage of year plant is operating
+Electricity price                                 :math:`P`           ``electricity_cost``                   $0.07/kWh          Electricity price in 2018 USD
+Electricity carbon intensity                      :math:`f_{eci}`     ``electrical_carbon_intensity``        0.475 kg/kWh       Carbon intensity of electricity
+Capital recovery factor                           :math:`f_{crf}`     ``capital_recovery_factor``            None               Calculated by derived class 
+=============================================  ====================  =====================================  ===============  ==============================================================================
+
+
+Aggregates calculation
```
Comment (adam-a-a) [inline docs/technical_reference/costing/costing_base.rst]: "Aggregate Metrics"

File: docs/technical_reference/costing/watertap_costing.rst [hunk 17]
```diff
@@ -0,0 +1,168 @@
+.. _watertap_costing:
+
+WaterTAP Costing Package
+========================
+
+.. currentmodule:: watertap.costing.watertap_costing
+
+The WaterTAP costing module contains the costing package typically used for high-fidelity models, but it can also be utilized with zero-order models. Technoecomic data is set on the package or utilizing global parameters added by the unit model costing method.
```
Comment (adam-a-a) [inline docs/technical_reference/costing/watertap_costing.rst]: "Technoeconomic data"

File: docs/technical_reference/costing/watertap_costing.rst [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+=============================================  ====================  ===============================================  ==============================================================================
+
+Calculations for each of these costs are presented below.
+
+Costing Index and Technoeconomic Factors
+----------------------------------------
+
+Default costing indices are provide with the WaterTAP Costing Package, but the user is free to modify these for their needs.
+
+WaterTAP uses the Consumer Price Index (CPI) to help account for the time-value of investments and are used in the capital
```
Comment (adam-a-a) [inline docs/technical_reference/costing/watertap_costing.rst]: Hmm.. I thought we used the Chemical Engineering Plant Cost Index (CEPCI)

File: watertap/costing/tests/test_costing_interop.py [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+    m.fs.costing.LCOW_with_revenue = pyo.Expression(
+        expr=(
+            (
+                m.fs.costing.total_annualized_cost
+                + m.fs.costing.aggregate_flow_costs["ferric_chloride"]
+            )
+            / m.fs.costing.annual_water_production
+        ),
+        doc="Levelized Cost of Water With Revenue",
+    )
```
Comment (adam-a-a) [inline watertap/costing/tests/test_costing_interop.py:80]: For the future, I wonder if we want to build in revenue as a separate aggregate that can be switched on or off as desired and incorporated (or excluded) from LCOW calculations, without have to do so manually like so.

**Review by adam-a-a — COMMENTED** [review]
This looks great. I have some minor suggestions (mostly typo fixes) and questions on parts that I just wasn't clearly understanding (even though I assume they work as intended). 

EDIT: I also had several comments on things we might want to consider in the future.

**Review by adam-a-a — APPROVED** [review]
@bknueven I am going to go ahead and approve this so that we can merge ASAP after typos addressed.

File: docs/technical_reference/costing/watertap_costing.rst (hunk 18 above)
Comment (bknueven) [reply docs/technical_reference/costing/watertap_costing.rst]: You are correct. I copied this sentence from the ZO costing documentation and fixed it in both places.

File: watertap/core/util/misc.py (hunk 6 above)
Comment (bknueven) [reply watertap/core/util/misc.py:31]: Yes, more or less correct. This function checks to see if there are any non-constant terms (e.g., Vars, mutable Params) in the expression. Unfortunately, Pyomo's existing checks methods do not cover this use case.

As for the purpose, I will direct you this comment: https://github.com/watertap-org/watertap/pull/1122#discussion_r1320341507

File: watertap/core/util/tests/test_misc.py (hunk 7 above)
Comment (bknueven) [reply watertap/core/util/tests/test_misc.py:46]: Yes.

File: watertap/core/util/tests/test_misc.py (hunk 8 above)
Comment (bknueven) [reply watertap/core/util/tests/test_misc.py:55]: Actually, this is a good question. I have to admit that I wrote this test several months ago.

Comment (bknueven) [reply watertap/core/util/tests/test_misc.py:55]: Apparently so. In fact, you can't set a `Param` with units to be *not* mutable:
```
In [9]: m.param_units_2 = Param(initialize=42, units=units.m**2 / units.s, mutable=False )

In [10]: m.param_units_2.mutable
Out[10]: True
```

This seems like a bug. I've created an issue in Pyomo (https://github.com/Pyomo/pyomo/issues/3002).

Comment (andrewlee94) [reply watertap/core/util/tests/test_misc.py:55]: @bknueven No, this is not a bug. Due to how `Params` work, if they are given units then they must be mutable.

File: watertap/costing/costing_base.py (hunk 13 above)
Comment (bknueven) [reply watertap/costing/costing_base.py:174]: They both still exist -- would need further refinement to eliminate one or the other.

**Comment by bknueven** [comment]
Thanks for the detailed comments @adam-a-a. I have captured them in #1147 so we don't lose track of them.
